src/impl/wone/kernels.cpp

```diff
@@ -0,0 +1,71 @@
+#include "wone.hpp"
+
+namespace wash {
+
+    void ForceKernel::exec() const {
+#pragma omp parallel for
+        for (auto& p : get_particles()) {
+            // TODO: perhaps remove neighbors from ForceFuncT, since neighbors can be accessed directly from a particle
+            const std::vector<Particle> neighbours = p.get_neighbors();
+            func(p, neighbours);
+        }
+    }
+
+    void UpdateKernel::exec() const {
+#pragma omp parallel for
+        for (auto& p : get_particles()) {
+            func(p);
+        }
+    }
+
+    void ReductionKernel::exec() const {
+        double local_result;
+        MPI_Op mpi_op;
+        switch (reduce_op) {
+        case ReduceOp::max:
+            local_result = -std::numeric_limits<double>::infinity();
+#pragma omp parallel for reduction(max : local_result)
+            for (auto& p : get_particles()) {
+                local_result = std::max(local_result, map_func(p));
+            }
+            mpi_op = MPI_MAX;
+            break;
+
+        case ReduceOp::min:
+            local_result = std::numeric_limits<double>::infinity();
+#pragma omp parallel for reduction(min : local_result)
+            for (auto& p : get_particles()) {
+                local_result = std::min(local_result, map_func(p));
+            }
+            mpi_op = MPI_MIN;
+            break;
```

Comment:
The `kernels.cpp` file is duplicated for each backend. I'm wondering if we can isolate the common functionality into one file, then have the implementation dependent code e.g. the `AllReduce_` for each backend. Again, this is just a nitpick though